README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
ABHgenotypeR
============

The ABHgenotypeR R package to easily visualize and manipulate ABH genotypes.

Introduction
------------

The `ABHgenotypeR` package provides simple imputation, error-correction and plotting capacities for genotype data. The function in this package were initially developed for the GBS/QTL analysis pipeline described in:

*Furuta, Reuscher et. al., 2016 Adaption genotyping by sequencing for rice F2 populations. BMC Genomics XYZ*

The `ABHgenotypeR` package is supposed to serve as an intermediate but independent analysis tool between the TASSEL GBS pipeline and the `qtl` package. `ABHgenotypeR` provides functionalities not found in either TASSEL or `qtl` in addition to visualization of genotypes as "graphical genotypes". Plotting is based on ggplot2.

Installation
------------

Note:
=====

CRAN V1.0.0 contains a bug which missorts markers, install from git for now

from CRAN:

``` r
install.package("ABHgenotypeR")
library(ABHgenotypeR)
```

or from github:

``` r
#needs "devtools"
install.packages("devtools")
library(devtools)

install_github("StefanReuscher/ABHgenotypeR", build_vignettes = TRUE)
```

for more information run:

``` r
vignette("ABHgenotypeR-vignette", package = "ABHgenotypeR")
```


NEWS.md
# ABHgenotypeR V1.0.1
 fixed a bug were plot functions would sort markers wrongly

# ABHgenotypeR V1.0.0
Release version


R/imputeByFlanks.R
#' Impute missing genotypes based on flanking alleles
#'
#' @param inputGenos A genotypes list object.
#'
#' @return A genotype object in which missing data is imputed based on flanking
#'   alleles. Any number of N is replaced by either A, B or N if the alleles which flank the N match
#'
#' @examples \dontrun{imputedGenos <- imputeByFlanks(genotypes)}

#' @export
imputeByFlanks <- function (inputGenos = "genotypes") {

  geno_raw <- inputGenos$ABHmatrix
  #setup empty matrix for imputed genotypes
  geno_imp <- matrix(0,
                     nrow = nrow(geno_raw),
                     ncol = ncol(geno_raw))

  #loop through chromosome
  for(chrom_count in 1:12) {
    geno_temp <- geno_raw[,inputGenos$chrom == chrom_count]
    #loop through rows, replace Ns flanked by the same parent/het
    for (row_count in 1:nrow(geno_temp)){
      geno_imp[row_count, inputGenos$chrom == chrom_count] <- substring(gsub(x = paste(geno_temp[row_count,],
                                                                            collapse = ""),
                                                                  pattern = "(?:A|(?<!^)\\G)\\KN(?=N*A)",
                                                                  replacement = "A",
                                                                  perl = TRUE),
                                                             1:ncol(geno_temp),
                                                             1:ncol(geno_temp))
      geno_imp[row_count, inputGenos$chrom == chrom_count] <- substring(gsub(x = paste(geno_imp[row_count, inputGenos$chrom == chrom_count],
                                                                            collapse = ""),
                                                                  pattern = "(?:B|(?<!^)\\G)\\KN(?=N*B)",
                                                                  replacement = "B",
                                                                  perl = TRUE),
                                                             1:ncol(geno_temp),
                                                             1:ncol(geno_temp))
      geno_imp[row_count, inputGenos$chrom == chrom_count] <- substring(gsub(x = paste(geno_imp[row_count, inputGenos$chrom == chrom_count],
                                                                            collapse = ""),
                                                                  pattern = "(?:H|(?<!^)\\G)\\KN(?=N*H)",
                                                                  replacement = "H",
                                                                  perl = TRUE),
                                                             1:ncol(geno_temp),
                                                             1:ncol(geno_temp))
    }
  }
  outputGenos <-inputGenos
  outputGenos$ABHmatrix <- geno_imp
  dimnames(outputGenos$ABHmatrix) <- list("individual_names" = inputGenos$individual_names,
                                        "marker_names" = inputGenos$marker_names)
  reportGenos(inputGenos)
  cat(paste("\n"))
  reportGenos(outputGenos)

  outputGenos
}


R/correctStretches.R
#' Correct short miscalled stretches based on flanking alleles.
#'
#' @param inputGenos A genotypes list object.
#' @param maxHapLength The maximum length of stretches flanked
#'   by non-heterzygous sites that are changed. If set to 1
#'   (default) only AXA or BXB will be corrected. If set to 2, both AXA and AXYA
#'   (or BXB and BXYB) will be corrected.
#'
#' @return A genotype object in which short miscalled stretches are
#'   corrected if both flanking alleles match.
#'
#' @examples \dontrun{corrStretchGenos <- correctStretches(genotypes, maxHapLength = 3)}

#' @export
correctStretches <- function(inputGenos = "genotypes",
                             maxHapLength = 1) {

  geno_raw <- inputGenos$ABHmatrix
  #setup a matrix for correcting errors
  geno_correctedErr <- matrix(0,
                              nrow = nrow(geno_raw),
                              ncol = ncol(geno_raw))
  #make reg expressions for A
  patExprA <- NULL # a character vector which holds regexp
  for(i in 1:maxHapLength) {
    patExprA[i] <- paste("(A)([BHN]{",i,"})(?=A)", sep = "")
  }
  replExprA <- NULL # a character vector which holds regexp
  for(i in 1:maxHapLength) {
    replExprA[i] <- paste("\\1",
                          paste(rep("\\1",i), sep = "", collapse = ""),
                          sep = "", collapse = "")
  }

  #make reg expressions for B
  patExprB <- NULL # a character vector which holds regexp
  for(i in 1:maxHapLength) {
    patExprB[i] <- paste("(B)([AHN]{",i,"})(?=B)", sep = "")
  }
  replExprB <- NULL # a character vector which holds regexp
  for(i in 1:maxHapLength) {
    replExprB[i] <- paste("\\1",
                          paste(rep("\\1",i), sep = "", collapse = ""),
                          sep = "", collapse = "")
  }

  #in geno_raw replace errors
  for(chrom_count in 1:12) {
    geno_temp <- geno_raw[,inputGenos$chrom == chrom_count]
    for (row_count in 1:nrow(geno_correctedErr)) {
      for(HapLen in 1:length(patExprA)) { #replace with A
        if(HapLen == 1) { #first replace in genotemp, then in geno_correc
          geno_correctedErr[row_count, inputGenos$chrom == chrom_count] <- substring(gsub(x = paste(geno_temp[row_count,],
                                                                                                    collapse = ""),
                                                                                          pattern = patExprA[HapLen],
                                                                                          replacement = replExprA[HapLen],
                                                                                          perl = TRUE),
                                                                                     1:ncol(geno_temp),
                                                                                     1:ncol(geno_temp))
        } else {
          geno_correctedErr[row_count, inputGenos$chrom == chrom_count] <- substring(gsub(x = paste(geno_correctedErr[row_count, inputGenos$chrom == chrom_count],
                                                                                                    collapse = ""),
                                                                                          pattern = patExprA[HapLen],
                                                                                          replacement = replExprA[HapLen],
                                                                                          perl = TRUE),
                                                                                     1:ncol(geno_temp),
                                                                                     1:ncol(geno_temp))
        }
      }
      for(HapLen in 1:length(patExprB)) { #replace with B, only in geno_correctedErr
        geno_correctedErr[row_count, inputGenos$chrom == chrom_count] <- substring(gsub(x = paste(geno_correctedErr[row_count, inputGenos$chrom == chrom_count],
                                                                                                  collapse = ""),
                                                                                        pattern = patExprB[HapLen],
                                                                                        replacement = replExprB[HapLen],
                                                                                        perl = TRUE),
                                                                                   1:ncol(geno_temp),
                                                                                   1:ncol(geno_temp))
      }
    }
  }

  outputGenos <-inputGenos
  outputGenos$ABHmatrix <- geno_correctedErr
  dimnames(outputGenos$ABHmatrix) <- list("individual_names" = inputGenos$individual_names,
                                          "marker_names" = inputGenos$marker_names)

  reportGenos(inputGenos)
  cat(paste("\n"))
  reportGenos(outputGenos)

  outputGenos
}



R/plotCompareGenos.R
#' Compare to genotype matrices
#'
#' @param genos_1 Output of readABHgenotypes
#' @param genos_2 Output of readABHgenotypes. Note that both genos object need to
#' have identical numbers of marker x individuals.
#' @param markerToPlot A character vector of marker names which appear in the
#'   plot. Defaults to all.
#' @param individualsToPlot A character vector of individual names which appear
#'   in the plot. Defaults to all.
#' @param chromToPlot A character vector of chromosome names which appear in the
#'   plot. Defaults to all.
#' @param CompColors A character vector of length 2 giving the color names or
#'   values to use for differnt and identical markers.
#'   Defaults to black and orange.
#' @param textSize The size of all text elements in the plot. Useful for making a
#'  nice plot. Defaults to 12.
#' @param showMarkerNames Show the marker names along the x axis. This and
#'  showIndividualnames are useful when you display only a few markers and
#'  want them labeled. Defaults to FALSE.
#' @param showIndividualNames Show individual names along the y axis.
#'
#' @return A graphical comparison of genotypes.
#'
#' @examples \dontrun{plotCompareGenos(preImpGenotypes,postImpGenotypes)}
#' \dontrun{#for more examples see plotGenos()}

#' @export
plotCompareGenos <- function(genos_1 = "genotypes_1",
                             genos_2 = "genotypes_2",
                             markerToPlot = "all",
                             individualsToPlot = "all",
                             chromToPlot = "all",
                             CompColors = c("#000000", "#E69F00"),
                             textSize = 12,
                             showMarkerNames = FALSE,
                             showIndividualNames = FALSE) {

  if(showMarkerNames == TRUE) {textX <- element_text(colour = "black", angle = 90)}
  else{textX <- element_blank()}

  if(showIndividualNames == TRUE) {textY <- element_text(colour = "black")}
  else{textY <- element_blank()}

  comp <- genos_1$ABHmatrix == genos_2$ABHmatrix

  comp_df <- as.data.frame(t(comp*1)) #logical -> numerical -> df
  comp_df$chrom <- genos_1$chrom
  comp_df$marker_names <- genos_1$marker_names

  comp_df <- reshape2::melt(comp_df,
                            id.vars = c("marker_names","chrom"),
                            variable.name = "individual_names",
                            value.name = "comp")

  if(markerToPlot[1] != "all") comp_df <- comp_df[comp_df$marker_names %in% markerToPlot,]

  if(individualsToPlot[1] != "all") comp_df <- comp_df[comp_df$individual_names %in% individualsToPlot,]

  if(chromToPlot[1] != "all") comp_df <- comp_df[comp_df$chrom %in% chromToPlot,]

  comp_df$individual_names <- factor(comp_df$individual_names,
                                     levels = rev(levels(comp_df$individual_names)))

  marker_names <- individual_names <- comp <- NULL #appease R cmd check

  ggplot(comp_df, aes(x = marker_names, y = individual_names,  fill = factor(comp)))+
    geom_tile()+
    scale_fill_manual(values = CompColors,
                      name = "genotype comparison",
                      labels = c("different","identical"))+
    facet_grid(.~chrom, scales = "free", space = "free_x")+
    ylab("individuals")+
    xlab("markers")+
    theme(text = element_text(size = textSize),
          axis.text.x = textX,
          axis.text.y = textY,
          axis.ticks = element_blank(),
          panel.background = element_blank(),
          legend.position = "bottom")
}


R/correctUndercalledHets.R
#' Correct undercalled heterozygous sites based on flanking alleles.
#'
#' @param inputGenos A genotypes list object.
#' @param maxHapLength The maximum length of not heterozygous stretches flanked
#'   by heterzygous sites that are changed to heterozygous. If set to 1
#'   (default) only HAH or HBH will be corrected. If set to 2, both HAH and HAAH
#'   (or HBH and HBBH) will be corrected.
#'
#' @return A genotype object in which undercalled heterozygous sites are
#'   corrected if both flanking alleles match.
#'
#' @examples \dontrun{corrUndHetsGenos <- correctUndercalledHets(genotypes, maxHapLength = 3)}

#' @export
correctUndercalledHets <- function(inputGenos = "genotypes",
                                   maxHapLength = 1) {

  geno_raw <- inputGenos$ABHmatrix
  #setup a matrix for correcting undercalled hets
  geno_correctedHets <- matrix(0,
                               nrow = nrow(geno_raw),
                               ncol = ncol(geno_raw))
  #make reg expressions for H
  patExprH <- NULL # a character vector which holds regexp
  for(i in 1:maxHapLength) {
    patExprH[i] <- paste("(H)([ABN]{",i,"})(?=H)", sep = "")
  }
  replExprH <- NULL # a character vector which holds regexp
  for(i in 1:maxHapLength) {
    replExprH[i] <- paste("\\1",
                          paste(rep("\\1",i), sep = "", collapse = ""),
                          sep = "", collapse = "")
  }


  #in geno_raw replace undercalled hets
  for(chrom_count in 1:12) {
    geno_temp <- geno_raw[,inputGenos$chrom == chrom_count]
    for (row_count in 1:nrow(geno_correctedHets)) { #first replace in genotemp, then in geno_corrected
      for(HapLen in 1:length(patExprH)) {
        if(HapLen == 1) {
          geno_correctedHets[row_count, inputGenos$chrom == chrom_count] <- substring(gsub(x = paste(geno_temp[row_count,],
                                                                                          collapse = ""),
                                                                                pattern = patExprH[HapLen],
                                                                                replacement = replExprH[HapLen],
                                                                                perl = TRUE),
                                                                           1:ncol(geno_temp),
                                                                           1:ncol(geno_temp))
        } else {
          geno_correctedHets[row_count, inputGenos$chrom == chrom_count] <- substring(gsub(x = paste(geno_correctedHets[row_count, inputGenos$chrom == chrom_count],
                                                                                          collapse = ""),
                                                                                pattern = patExprH[HapLen],
                                                                                replacement = replExprH[HapLen],
                                                                                perl = TRUE),
                                                                           1:ncol(geno_temp),
                                                                           1:ncol(geno_temp))
        }
      }
    }
  }
#put feedbackfunction here
  outputGenos <-inputGenos
  outputGenos$ABHmatrix <- geno_correctedHets
  dimnames(outputGenos$ABHmatrix) <- list("individual_names" = inputGenos$individual_names,
                                       "marker_names" = inputGenos$marker_names)

  reportGenos(inputGenos)
  cat(paste("\n"))
  reportGenos(outputGenos)

  outputGenos
  }



R/plotGenos.R
#' Plot graphical genotypes.
#'
#' @param genos The output of readABHgenotypes
#' @param markerToPlot A character vector of marker names which appear in the
#'   plot. Defaults to all.
#' @param individualsToPlot A character vector of individual names which appear
#'   in the plot. Defaults to all.
#' @param chromToPlot A character vector of chromosome names which appear in the
#'   plot. Defaults to all.
#' @param alleleColors A character vector of length 4 giving the color names or
#'   values to use for the A,B,H and n.d genotypes. Defaults to orange, blue,
#'   green and black.
#' @param textSize The size of all text elements in the plot. Useful for making a
#'  nice plot. Defaults to 12.
#' @param showMarkerNames Show the marker names along the x axis. This and
#'  showIndividualNames are useful when you display only a few markers and
#'  want them labeled. Defaults to FALSE.
#' @param showIndividualNames Show individual names along the y axis.
#'
#' @return Graphical genotypes.
#'
#' @examples \dontrun{plotGenos(genotypes)}
#' markerNames <- c("marker1", "marker2", "marker3")
#' individualNames <- c("F2_100", "F2_101", "F2_102", "F2_103")
#' someColors <- c("black", "red", "gold", "white")
#' \dontrun{plotgenos(genotypes, markerNames, individualNames, 1:3, someColors)}
#'
#' \dontrun{p <- plotGenos(genotypes)}

#' @export
plotGenos <- function(genos = "genotypes",
                      markerToPlot = "all",
                      individualsToPlot = "all",
                      chromToPlot = "all",
                      alleleColors = c("#56B4E9","#E69F00",
                                       "#009E73", "#000000"),
                      textSize = 12,
                      showMarkerNames = FALSE,
                      showIndividualNames = FALSE) {

  if(showMarkerNames == TRUE) {textX <- element_text(colour = "black", angle = 90)}
  else{textX <- element_blank()}

  if(showIndividualNames == TRUE) {textY <- element_text(colour = "black")}
  else{textY <- element_blank()}

  ggt <- data.frame(t(genos$ABHmatrix), stringsAsFactors = FALSE, check.names = FALSE)

  ggt$chrom <- genos$chrom
  ggt$index <- 1:length(genos$chrom)
  ggt$marker_names <- genos$marker_names

  ggt <- reshape2::melt(ggt,
                        id.vars = c("chrom", "index","marker_names"),
                        variable.name = "individual_names", value.name = "allele")

  if(markerToPlot[1] != "all") ggt <- ggt[ggt$marker_names %in% markerToPlot,]

  if(individualsToPlot[1] != "all") ggt <- ggt[ggt$individual_names %in% individualsToPlot,]

  if(chromToPlot[1] != "all") ggt <- ggt[ggt$chrom %in% chromToPlot,]

  marker_names <- individual_names <- allele <- NULL #appease R cmd check

  ggplot(ggt, aes(x = marker_names, y = individual_names,
                  fill = allele))+
    geom_tile()+
    scale_fill_manual(name = "genotypes",
                      values = c("A" = alleleColors[1], "B" = alleleColors[2],
                                 "H" = alleleColors[3], "N" = alleleColors[4]),
                      labels = c(genos$nameA, genos$nameB, "hetero", "n.d."))+
    facet_grid(.~chrom, scales = "free", space = "free_x")+
    xlab("marker")+
    ylab("individuals")+
    theme(text = element_text(size = textSize),
          axis.text.x = textX,
          axis.text.y = textY,
          axis.ticks = element_blank(),
          panel.grid = element_blank(),
          panel.background = element_blank(),
          strip.background = element_blank(),
          legend.position = "bottom",
          panel.margin = unit(0, "lines"))

}
